Deep Dive: Prometheus (#45)
tikv/website: content/deep-dive/monitoring/prometheus.md

---
title: Prometheus
weight: 1
---

# Prometheus

[Prometheus] is the monitoring system used in TiKV. It is an open-source
systems monitoring and alerting toolkit. Widely adopted by cloud native
applications, Prometheus is a graduated project of the [CNCF].

> The TiKV team develop and maintain [rust-prometheus], the Rust clien for

Reviewer: ```suggestion
The TiKV team develops and maintains [rust-prometheus], a Rust client for
```

Prometheus. The library is heavily used in TiKV to collect various metrics. 

## Data Model

> Prometheus stores all data as time series. Time series data are formed with

Reviewer: ```suggestion
Prometheus stores all data as time series. Time series data are made up of
```

_samples_. Each sample consists of a float64 value and a millisecond-precision
timestamp.

Every time series is uniquely identified by its _metric name_ and a set of
_labels_. Given a metric name and a set of labels, time series are frequently
identified using this notation:

```
<metric name>{<label name>=<label value>, ...}
```

## Metrics

Prometheus offers four [metric types][mt]: Counter, Gauge, Histogram and
Summary. The first three types of metrics are supported by rust-prometheus.

Next, let us see how we use these metrics in TiKV for statistics and issue
diagnosis.

### Counter

> A counter stores the totality of some value at each point in time. Therefore,

Reviewer: Can we note it is monotonically increasing?

vector. Finally, we use `sum()` because there could be data of various
instances and databases.

### Gauge

Unlike a counter, a gauge does not require the value to monotonically increase.
It can store time series of arbitrary numerical values.

In TiKV, the number of scheduler pending commands, which can go both up and
down, is represented by the [`tikv_scheduler_contex_total` gauge][sg]. Its
value is increased when a command is [enqueued][enq] and is decreased when a
command is [dequeued][deq] from the scheduler.

The total number of pending commands can be queried by:

```
sum(tikv_scheduler_contex_total{instance=~"$instance"}) by (instance)
```

### Histogram

A histogram is composed of multiple counters: the sum of all observed values,
the count of observed events, and some cumulative counters of configurable
buckets.

At the creation of a histogram, you need to set the upper bound of each bucket.
Every time a value is observed, the counters of the matching buckets are
incresed.

For example, the upper bounds of the buckets are `0.05`, `0.2`, `1` and `5`.
If a value `0.5` is observed, the value will be added to the sum, the total

> event count will be increased, and the counters of buckets `(-∞, 1]`, `(-∞, 5]`

Reviewer: I think this example is a bit unclear. :(
